### ISSUE TITLE
Focus does not converge and results in NaN.

### ISSUE THREAD

**Issue by gesen2egee** [LoganBooker/prodigy-plus-schedule-free#13]
Using the new update, Focus explodes after just a few hundred steps.

![Image](https://github.com/user-attachments/assets/49693bd0-56e9-4b5a-b131-969f35dad3bb)

In fact, I have tested a version locally migrated by pytorch optimizer , and it can converge while slightly outperforming the original Adam version.
```
            elif group['use_focus']:                
                # FOCUS: First Order Concentrated Updating Scheme: https://arxiv.org/pdf/2501.12243
                gamma = 0.2 # 0.2 is used more consistently in the paper, 0.1 is optimiser default.
                exp_avg, pbar = state['exp_avg'], state['pbar']
                exp_avg.mul_(beta1).add_(grad, alpha=1.0 - beta1)
                pbar.mul_(beta2).add_(p, alpha=1.0 - beta2)
                bias_correction2 = 1.0 - beta2 ** k
                pbar_hat = pbar / bias_correction2                
                update = (p - pbar_hat).sign_().mul_(gamma).add_(torch.sign(exp_avg))
```

Additionally, I have tested orthogonal gradients, 
and the Prodigy steps do not increase the learning rate, 
leading to worse results.

**Comment by LoganBooker** [comment]
@gesen2egee Thanks for the report -- there was an error in the implementation where the denominator was still having the sqrt applied as though it was a regular Adam second moment. Should be fixed in https://github.com/LoganBooker/prodigy-plus-schedule-free/commit/859ecfb9bec95fb4bfb6cc4624af1f87e77da20a. Even so, we're still applying Focus with Prodigy and Schedule Free, so the results might not directly translate.

As for OrthoGrad -- I'll have to investigate that further. I haven't observed any issues locally with the LR not increasing, but I do know the technique is not a silver bullet, it can definitely just be straight up worse for some scenarios.

**Comment by gesen2egee** [comment]
It’s possible that I’m training Stable Diffusion, which might not be very suitable.

Thank you for your work!

**Comment by gesen2egee** [comment]
@LoganBooker 

Change
self.update_prodigy(state, group, p.grad, p)
to
self.update_prodigy(state, group, grad, p)

orthograd can then increase the learning rate.
Perhaps the evaluation of grad needs to be consistent.

**Comment by LoganBooker** [comment]
Hmm, interesting. By that stage, `grad` is actually `update` due to the in-place operations the optimiser does, so you're actually passing the post-Adam update to Prodigy, rather than the original gradient.

It might be more correct to do (at least to test):
```
grad = self.orthograd(z_state, p.grad) if group['use_orthograd'] else p.grad.to(dtype=torch.float32, copy=True)
grad_copy = grad.clone()
```
Then later:
```
self.update_prodigy(state, group, grad_copy, y)
```
Which would feed in the OrthoGrad gradient. Though my intuition (which admittedly could be wrong) would be to use the raw gradient rather than modified one.

Alternatively, you could try `use_speed=True` instead.
